fix idris with megaparsec override

Jake-Gillberg opened the pull request:
<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!
List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

###### Motivation for this change
Idris 1.3.3 is broken

https://github.com/NixOS/nixpkgs/issues/86500

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [x] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

Jake-Gillberg committed:
fix idris with megaparsec override

Jake-Gillberg commented:
@cdepillabout, thanks for the direction. Is there a way I can manually run the re-generation script? When testing this, I'm still getting that megaparsec_8_0_0 is missing.

sternenseemann commented:
> @cdepillabout, thanks for the direction. Is there a way I can manually run the re-generation script? When testing this, I'm still getting that megaparsec_8_0_0 is missing.

Running `hackage2nix` is explained here: <https://github.com/NixOS/cabal2nix/blob/master/hackage2nix/README.md>

Jake-Gillberg commented:
Hmm, I'm afraid I'll have to have this spelled out for me. Here is what I have tried:

```
bash-4.4$ stack --nix build
Error parsing targets: The specified targets matched no packages.
Perhaps you need to run 'stack init'?
```
```
bash-4.4$ stack --nix build idris

Error: While constructing the build plan, the following exceptions were encountered:

In the dependencies for idris-1.3.3:
    Cabal-3.2.1.0 from stack configuration does not match >=2.4 && <3.1  (latest matching version is 3.0.2.0)
    aeson-1.5.5.1 from stack configuration does not match >=0.6 && <1.5  (latest matching version is 1.4.7.1)
    haskeline-0.8.1.1 from stack configuration does not match ==0.7.*  (latest matching version is 0.7.5.0)
    megaparsec-9.0.1 from stack configuration does not match >=7.0.4 && <9  (latest matching version is 8.0.0)
needed since idris is a build target.

Some different approaches to resolving this:

  * Set 'allow-newer: true'
    in /home/jgillberg/.stack/config.yaml to ignore all version constraints and build anyway.

  * Recommended action: try adding the following to your extra-deps
    in /home/jgillberg/.stack/global-project/stack.yaml:

- Cabal-3.0.2.0@sha256:f45ee078a8529d1ad3993767345ac83304c7f2ff5eec02c455a1dc200ad22c67,26714
- aeson-1.4.7.1@sha256:6d8d2fd959b7122a1df9389cf4eca30420a053d67289f92cdc0dbc0dab3530ba,7098
- haskeline-0.7.5.0@sha256:2d7f14c0d2e4054c0628216c2ed8e7de9f9c0b6f55f7026d5fe26f7726f40e45,4833
- megaparsec-8.0.0@sha256:362f85e243ecbcb550e1de6e5c74ba5a50c09edaa1208c99bc5b9fd242227fc6,3808

Plan construction failed.
```

sternenseemann commented:
Okay, seems like the instructions are outdated, sorry, I've never tried this myself actually. I'll try to make a PR updating the instructions later.

I managed to get this working like this:

```
git clone https://github.com/NixOS/cabal2nix.git
cd cabal2nix

ln -s $HOME/src/nix/nixpkgs nixpkgs # my local nixpkgs
cd nixpkgs
gh checkout pr 114005  # get your pr
cd ..

git clone https://github.com/commercialhaskell/all-cabal-hashes.git --branch hackage hackage

cabal2nix --shell . > shell.nix # get a dev environment with cabal2nix installled from nixpkgs
nix-shell
nix-shell> cabal v2-build
nix-shell> ./update-nixpkgs.sh # push fails for me because it's your branch
cd nixpkgs
env NIXPKGS_ALLOW_BROKEN=1 nix-build -A haskellPackages.idris # works!
```

Note also that I had to use `NIXPKGS_ALLOW_BROKEN=1` because `idris` is still marked as broken in `configuration-hackage2nix.yaml`. This can easily be fixed, you just have to remove idris from the appropriate list in there!

PS: as of writing this idris hasn't compiled yet, but I think no resolver errors is a positive sign :)

sternenseemann commented:
See also <https://github.com/NixOS/cabal2nix/pull/483>.

sternenseemann committed:
haskellPackages.idris: add patch for haskeline >= 0.8

0.8.0.0 removes the System.Console.Haskeline.MonadException module idris
1.3.3 depends on. A patch from master we pull in fixes the ensuing
compilation failure.

sternenseemann committed:
haskellPackages.idris: mark as unbroken

sternenseemann committed:
hackage-packages.nix: automatic Haskell package set update

This update was generated by hackage2nix v2.16.0-14-g12ea7a2 from Hackage revision
https://github.com/commercialhaskell/all-cabal-hashes/commit/f5a5ed7106a71c32d88da77b9c41257862ba2f07.

sternenseemann commented:
I've figured out the [remaining fixes on my branch](https://github.com/Jake-Gillberg/nixpkgs/compare/patch-idris...sternenseemann:patch-idris) feel free to pull in the commits :)

sternenseemann commented:
Seems like on `master` `idris` has been adapted to `megaparsec` 9: <https://github.com/idris-lang/Idris-dev/commit/6ea9bc913877d765048d7cdb7fc5aec60b196fac>

I'd be in favor of adding this patch as well instead of having an old version of `megaparsec` just for `idris` in `haskellPackages`.

Jake-Gillberg committed:
Merge pull request #1 from sternenseemann/patch-idris

Patch idris

sternenseemann commented:
Feel free to review the alternative solution with the `megaparsec` patch from master :) #114042

Jake-Gillberg committed:
alternative method- apply patches

Jake-Gillberg committed:
..

Jake-Gillberg commented:
Wow, thanks @sternenseemann, you are fast!

You must have a better method for getting those patch hashes than me..

sternenseemann commented:
> You must have a better method for getting those patch hashes than me..

Unfortunately you can't use `nix-prefetch-url` with `fetchpatch`, but you can use `sha256 = lib.fakeSha256;` and then the build wiill fail, but at least tell you the right hash.

`lib.fakeSha256` is actually just a shortcut for writing `"0000000000000000000000000000000000000000000000000000"` which is useful when determining hashes for all sorts of things in nixpkgs.

Jake-Gillberg committed:
..

Jake-Gillberg committed:
i'm learning

Jake-Gillberg committed:
hackage-packages.nix: automatic Haskell package set update

This update was generated by hackage2nix v2.16.0-13-gefcfbb2 from Hackage revision
https://github.com/commercialhaskell/all-cabal-hashes/commit/fc1c73689d6a4467ad1311c1161c95c9476d6d5e.

Jake-Gillberg commented:
Am I supposed to submit the updated `hackage-packages.nix`?

sternenseemann commented:
No, this will be regenerated on friday by peti anyways before `haskell-updates` is merged into master, so there is no need to.

Jake-Gillberg commented:
Okay, tested and the idris binary builds and works!

cdepillabout commented:
Superseded by https://github.com/NixOS/nixpkgs/pull/114042
